zcash_client_sqlite/src/wallet.rs

```diff
@@ -1969,6 +1970,89 @@ pub(crate) fn store_transaction_to_be_sent<P: consensus::Parameters>(
+        TransactionStatus::TxidNotRecognized | TransactionStatus::Mempool => {
+            // The transaction is unmined, so it either needs to be resubmitted if it is unexpired
+            // as of the current chain tip, or it needs to be marked as confirmed unmined as of the
+            // chain tip height. If the chain tip height is unknown, we're in a weird ignorable
+            // state; we'll just wait for the transaction status to be queried again later.
+            if let Some(chain_tip) = chain_tip_height(conn)? {
+                let mut stmt = conn.prepare_cached(
+                    "UPDATE transactions
+                     SET confirmed_unmined = :chain_tip
+                     WHERE txid = :txid
+                     RETURNING expiry_height",
+                )?;
+
+                let expiry_height = stmt
+                    .query_row(
+                        named_params![
+                            ":chain_tip": u32::from(chain_tip),
+                            ":txid": txid.as_ref()
+                        ],
+                        |row| {
+                            row.get::<_, Option<u32>>(0)
+                                .map(|opt| opt.map(BlockHeight::from))
+                        },
+                    )
+                    .optional()?
+                    .flatten();
+
+                // If the transaction's expiry height + the confirmation depth is less than the
+                // chain tip height, we know it to be un-mineable and we can remove the reference
+                // to it from the transaction retrieval queue.
+                if expiry_height
+                    .iter()
+                    .any(|expiry| *expiry + VERIFY_LOOKAHEAD < chain_tip)
+                {
+                    conn.execute(
+                        "DELETE FROM tx_retrieval_queue WHERE txid = :txid",
+                        named_params![
+                            ":txid": txid.as_ref()
+                        ],
+                    )?;
+                }
+            }
+        }
+        TransactionStatus::Mined(height) => {
+            // The transaction has been mined, so we can set its mined height, associate it with
+            // the appropriate block, and remove it from the retrieval queue.
+            let sql_args = named_params![
+                ":txid": txid.as_ref(),
+                ":height": u32::from(height)
+            ];
+
+            conn.execute(
+                "UPDATE transactions
+                 SET mined_height = :height
+                 WHERE txid = :txid",
+                sql_args,
+            )?;
+
+            conn.execute(
+                "UPDATE transactions
+                 SET block = blocks.height
+                 FROM blocks
+                 WHERE txid = :txid
+                 AND blocks.height = :height",
+                sql_args,
+            )?;
+
+            conn.execute(
+                "DELETE FROM tx_retrieval_queue WHERE txid = :txid",
```

Comment:
This does not check `query_status`, which is fine while `txid` remains `UNIQUE`, but if the previous comment is implemented, this deletion (and the equivalent one for `notify_tx_retrieved`) need to be gated on matching `query_status`.

Reply:
`txid` will remain unique, and enhancement requests will subsume status requests.